Change how data-workspace activates project extensions

kisantia opened the pull request:
<!-- Thank you for submitting a Pull Request. Please:
* Read our Pull Request guidelines:
  https://github.com/Microsoft/azuredatastudio/wiki/How-to-Contribute#pull-requests.
* Associate an issue with the Pull Request.
* Ensure that the code is up-to-date with the `main` branch.
* Include a description of the proposed changes and how to test them.
-->

This PR fixes https://github.com/microsoft/azuredatastudio/issues/21393. This change is so that the vscode-mssql extension doesn't always get activated when sql-database-projects-vscode gets activated by data-workspace-vscode (vscode-mssql gets activated because it's a dependency).

Now the sql projects extension will only get activated by data-workspace if the workspace or folder has a .sqlproj file, otherwise the extension will only get activated when the database projects viewlet is opened or on the other commands specified in the package.json. Data workspace activates the extension if there's a .sqlproj file present so that it will show up in the project tree right away when the viewlet is navigated to.

vscode-mssql getting activated when the SQL Server viewlet is opened - not activated by data-workspace or sql projects:
![image](https://user-images.githubusercontent.com/31145923/209033298-57a7774c-3ecb-4cc0-969d-db0d50c3aea3.png)

sql-database-projects-vscode and vscode-mssql getting activated when the folder has a .sqlproj:
![image](https://user-images.githubusercontent.com/31145923/209033226-55e476a6-2b2d-43b8-ba37-a06aac857b47.png)

sql-database-projects-vscode and vscode-mssql getting activated when no folder open and database projects viewlet is opened:
![image](https://user-images.githubusercontent.com/31145923/209033248-bdd42b10-f656-41d5-a3aa-36bcda3063e4.png)

sql-database-projects-vscode and vscode-mssql getting activated when no folder open and the new project command is run:
![image](https://user-images.githubusercontent.com/31145923/209033357-190d7988-42d8-4587-a256-49da2b392515.png)

Charles-Gagnon commented on extensions/data-workspace/src/services/workspaceService.ts:
```diff
@@ -278,7 +278,18 @@ export class WorkspaceService implements IWorkspaceService {
 		try {
 			if (!extension.isActive) {
-				await extension.activate();
+				const projectTypes = extension.packageJSON.contributes && extension.packageJSON.contributes.projects as string[];
```
```suggestion
				const projectTypes = extension.packageJSON.contributes?.projects as string[] | undefined;
				if (!projectTypes) {
					return;
				}
```

Charles-Gagnon replied:
But also - this function only appears to be called by something that already is looking through the projects contribution. So you could probably reuse the value from that and pass it here? You'll need to handle the case when it's undefined though (activating all extensions)

kisantia replied:
yeah I thought about passing the projType from `ensureProjectProviderLoaded()`, but then I still would need to do this anyway for the undefined (activate all extensions) case, so it didn't seem like there was much benefit to doing that.

Charles-Gagnon reviewed (commented):
So I was looking through the code and I think you'll need to do some additional work - there's certain places that all extensions are activated like the new project dialog/quickpick which I believe will still require activating all the extensions even if there's not an existing project in the workspace.

Charles-Gagnon commented:
Oh nevermind, that's what you changed the activation events for. But a couple things : 

1. It's not clear that extensions contributing to this will need to have that activation event - are you planning on documenting that somewhere? Or I still think it'd be fine to have the data workspace extension activate all contributing extensions when new project is ran - that's a direct action being taken so isn't as heavy (as opposed to automatically activating on startup like it's doing now)
2. Are you sure that the extension is guaranteed to be fully activated by the time the new project flows request information from those extensions? They aren't waiting on extension activation anymore in those cases so I think you have a race condition there

kisantia commented:
Updated so that the data-workspace extension activates all project providing extensions when the new and open commands start and removed the activation events on those commands from the sql-database-projects extension. It makes sense to remove the burden of knowing to activate on those commands from the project providing extensions, whenever other project providing extensions get added.

Charles-Gagnon commented on extensions/data-workspace/src/services/workspaceService.ts:
```diff
@@ -265,20 +266,44 @@ export class WorkspaceService implements IWorkspaceService {
+
+					let projFilesInWorkspace: vscode.Uri[] = [];
+					for (const projType of projectTypes) {
+						(await vscode.workspace.findFiles(`**/*.${projType}`)).forEach(f => projFilesInWorkspace.push(f));
```
This doesn't currently handle a project being added to the workspace after initial load - for example if they create it with some other extension or outside of VS Code itself. You'll need to do something like set up a file system watcher to handle that.

kisantia replied:
That should get handled by this: https://github.com/microsoft/azuredatastudio/blob/abc2ba41fc03fcea179976eef40ccca6950ed209/extensions/data-workspace/src/main.ts#L35-L37

The `workspaceTreeDataProvider.refresh()` calls `getProjectsInWorkspace()`, which then eventually calls into `handleProjectProviderExtension()` which will activate the project extension if the workspace now has a file with it's contributed project type.

Charles-Gagnon replied:
Sort of - but that only handles if a new folder is added. What if the user creates a project in a folder that's already opened?

kisantia replied:
Like if they create the project outside of ADS/vscode? The projects extension would get activated when they first go to the database projects viewlet, but they'd have to use the refresh button for the new project to show up the first time if the extension wasn't already activated. Not sure how common of a scenario this would be though, so it might be unnecessary overhead to have a watcher on every folder in the workspace and check for project files every time there's a change.  

If they create the project in ADS/vscode, then the projects extension will get activated by the new/open/create project from db commands, which is how I expect most people to create or open projects, besides opening a folder that contains a project.

Charles-Gagnon replied:
Alright, as long as you're ok with the possible limitations there. I agree it seems like more of an edge case.

Charles-Gagnon commented on extensions/data-workspace/src/services/workspaceService.ts:
```diff
@@ -265,20 +266,44 @@ export class WorkspaceService implements IWorkspaceService {
+						return;
+					}
+
+					let projFilesInWorkspace: vscode.Uri[] = [];
+					for (const projType of projectTypes) {
+						(await vscode.workspace.findFiles(`**/*.${projType}`)).forEach(f => projFilesInWorkspace.push(f));
+					}
+
+					if (projFilesInWorkspace?.length > 0) {
+						// only try to activate the extension if the workspace has a project with that extension
+						await extension.activate();
+					} else {
+						return;
+					}
```
This can be simplified - you only need to find one project that matches to justify activating the extension and then can skip checking all the other supported types at that point. 

```suggestion
					for (const projType of projectTypes) {
						const projFilesInWorkspace = await vscode.workspace.findFiles(`**/*.${projType}`);
						if (projFilesInWorkspace.length > 0) {
							// only try to activate the extension if the workspace has at least one project with that extension
							await extension.activate();
							break;
						}
					}
```
